packages/fscommerce/src/Commerce/sessions/CommerceCookieSessionManager.web.ts

```diff
@@ -16,40 +22,108 @@ export default class CommerceCookieSessionManager extends CommerceSessionManager
     return true;
   }
 
-  async get(): Promise<SessionToken> {
-    let token: SessionToken | null = null;
-
-    try {
-      token = await this.options.sessionCookiesToToken();
-    } catch (e) {
-      /* let it fail sliently */
-    }
-
-    if (token) {
-      return token;
-    }
-
-    if (this.options.restoreCookies) {
-      try {
-        await this.options.restoreCookies();
-        token = await this.options.sessionCookiesToToken();
-      } catch (e) {
-        /* let it fail sliently */
+  /**
+   * Returns current token for the session.
+   *
+   * Only one invocation at a time is allowed. If it's called again
+   * it will have to wait until the previous invocation is complete.
+   *
+   * The token is cached locally and re-used until it expires.
+   *
+   * The token is refreshed automatically upon expiration if `refreshToken` method was provided.
+   *
+   * If no token exist it will attempt to convert current session into JWT,
+   * but only if the cookie `dwanonymous_*` is present.
+   * Otherwise, it will create new token for the guest user.
+   *
+   * @returns {Promise.<SessionToken | null>} A Promise representing token
+   */
+  // tslint:disable:cyclomatic-complexity
+  async get(): Promise<SessionToken | null> {
+    // allow no more than one invocation at a time
+    this.getPromise = (this.getPromise || Promise.resolve()).then(async () => {
+      let token: SessionToken | null | undefined = this.token;
+
+      // we have expired token stored locally
+      if (token && this.isExpired(token)) {
+        // refresh it
+        token = await this.options.refreshToken(token);
+
+        // got new token - store it
+        if (token) {
+          await this.set(token);
+        }
       }
 
+      // got old token
       if (token) {
         return token;
       }
-    }
 
-    return this.options.createGuestToken();
+      // no token - get a token from the session
+      const sessionCookiesPresent = document.cookie?.includes('dwanonymous_');
+
+      if (sessionCookiesPresent) {
+        try {
+          token = await this.options.sessionCookiesToToken();
+        } catch (e) {
+          /* let it fail sliently */
+        }
+      }
+
+      // got a token - store it
+      if (token) {
+        await this.set(token);
+
+        return token;
+      }
+
+      // no token - try to restore the cookie and the a token from the session
+      if (this.options.restoreCookies) {
+        try {
+          await this.options.restoreCookies();
+
+          const sessionCookiesPresent = document.cookie?.includes('dwanonymous_');
+
+          if (sessionCookiesPresent) {
+            token = await this.options.sessionCookiesToToken();
+          }
+        } catch (e) {
+          /* let it fail sliently */
+        }
+      }
+
+      // got a token - store it
+      if (token) {
+        await this.set(token);
+
+        return token;
+      }
```

Comment:
I'm not entirely clear on the purpose of restoreCookies, but can we simplify this by doing the cookie restore first? eg.

```ts
// ...
      if (this.options.restoreCookies) {
        try {
          await this.options.restoreCookies();
        } catch (e) {
          /* let it fail sliently */
        }
      }

      // no token - get a token from the session
      const sessionCookiesPresent = document.cookie?.includes('dwanonymous_');
      if (sessionCookiesPresent) {
        try {
          token = await this.options.sessionCookiesToToken();
          // got a token - store it
          if (token) {
            await this.set(token);
            return token;
          }
        } catch (e) {
          /* let it fail sliently */
        }
      }
// ...
```

Reply:
@nathan-sankbeil,
I'm not aware of the users of the method `restoreCookies` and possible implementations.
I didn't touch current implementation and wary of doing that. I don't want to introduce unnecessary breaking changes.